src/os/bluestore/Compression.cc

```diff
@@ -82,6 +97,16 @@ inline bool Estimator::is_worth()
 {
   uint32_t cost = uncompressed_size * expected_compression_factor +
                   compressed_size * expected_recompression_error;
+  if (uncompressed_size == 0 && single_compressed_blob) {
+    // The special case if all extents are from compressed blobs.
+    // We want to avoid the case of recompressing into exactly the same.
+    // The cost should increase proportionally to blob size;
+    // the rationale is that recompressing small blob is likely to provide gain,
+    // but recompressing whole large blob isn't.
+    uint32_t padding_size = p2nphase(compressed_size, uint32_t(bluestore->min_alloc_size));
+    uint32_t split_tax = padding_size * compressed_area / target_blob_size;
```

Comment:
Just to verify: no chance of the 'padding_size * compressed_area' part of the computation exceeding 32b?